Connext

pedrouid opened the pull request.

pedrouid committed:
add connextlcient

pedrouid committed:
fix linting

pedrouid committed:
install @connext/client

pedrouid committed:
add initWallet and update init

pedrouid committed:
add ETH as a currency and replace Raiden with Connext on db references

pedrouid committed:
refactor error messages

pedrouid committed:
remove sendRequest from ConnextCLient

pedrouid committed:
fix trading limits

pedrouid committed:
change max amount

pedrouid committed:
cleanup

pedrouid committed:
polishing

pedrouid committed:
resolve conflicts

pedrouid committed:
fixes

pedrouid committed:
allow to override network with ethProviderUrl and nodeUrl

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,513 @@
+  }) {
+    super(logger, config.disable);
+
+    this.network = config.network || '';
```
Think the 3 values below should always be defined.

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,513 @@
+      nodeUrl: this.nodeUrl,
+      xpub: wallet.xpub,
+      keyGen: (index: string) => wallet.keyGen(index),
+      store: new ConnextStore(new FileStorage()),
```
If I'm reading the code correctly it will create a data directory to `./connext-store` in the runtime directory of xud? We could provide a path to xud's data directory here from config.

pedrouid replied:
Correct, this can be configured into the FileStorage class. Where would you like to assign the directory?

erkarl replied:
If we go with the REST API wrapper approach then just having it configurable via runtime arguments or config file would do.

pedrouid replied:
Correct, this is no longer relevant 👍

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,513 @@
+   */
+  private setTokenAddresses = (currencyInstances: CurrencyInstance[]) => {
+    currencyInstances.forEach((currency) => {
+      if (currency.tokenAddress) {
```
If we decide to go with 1 swap client per currency then this will also need a check for `SwapClient.Connext`.

kilrau replied:
> if we decide to go with 1 swap client per currency

definitely one swap client per currency, see below

sangaman replied:
On a side note (and not needed to be done in this PR) it would be good to have the same logic in RaidenClient, or actually better yet within SwapClientManager. I don't think this needs to be done as part of this PR though, I can take a look at implementing it separately.

pedrouid replied:
As per the other comment, WETH and DAI will use Raiden and ETH and USDT will use Connext

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,513 @@
+  }
+
+  public lookupPayment = async (
+    rHash: string,
```
`rHash` is unused.

pedrouid replied:
This will be revised once we have the HTLC functionality on our client

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,513 @@
+    currency?: string,
+    destination?: string,
+  ) => {
+    let endpoint = 'payments';
```
Lots of unused `endpoint` variables in this file. I'm assuming it's a copy paste leftover from `RaidenClient` and won't comment on it further.

erkarl commented on lib/constants/enums.ts:
```diff
@@ -2,6 +2,7 @@
 export enum SwapClientType {
   Lnd,
   Raiden,
+  Connext,
```
Now that we have more than 1 swap client we'll need to modify `SwapRecovery` L83 to also save the recovered preimages to memory. This is used when the payment succeeded, but xud managed to crash before using the preimage to claim the second payment.

pedrouid replied:
Is this still relevant if each asset only has one SwapClient?

erkarl commented on lib/db/seeds/mainnet.ts:
```diff
@@ -14,22 +14,29 @@ const nodes = [
   {
     id: 'WETH',
-    swapClient: SwapClientType.Raiden,
+    swapClient: SwapClientType.Raiden, // needs to be configurable to support Connext
```
I'm not quite sure if we want to have multiple swap clients per currency. Can you remind me what was the primary use case of having such functionality?

pedrouid replied:
The case here is that Connext and Raiden have an overlap of assets that they can swap. Meaning that if a user wants to swap either DAI or WETH it can be done either with Connext or Raiden. 

Alternatively you could duplicate the assets with different swapClients. Or just replace the Raiden client for DAI and  WETH

kilrau replied:
> Or just replace the Raiden client for DAI and WETH

That is the ultimate goal. And definitely only one swap client per asset. Until then we'd like to keep raiden in. Proposal how to do that: Leave WETH & DAI to Raiden. Give ETH & USDT to connext. Once everything is integrated, working and tested we'll also switch WETH & DAI to connext.

sangaman replied:
I'll take a look at better handling the option of more than 1 client for a given currency in a separate PR.

pedrouid replied:
Added USDT and assigned it SwapClientType.Connext

erkarl commented on lib/swaps/SwapClientManager.ts:
```diff
@@ -213,6 +226,11 @@ class SwapClientManager extends EventEmitter {
     }
 
+    if (this.connextClient) {
+      const initWalletPromise = this.connextClient.initWallet(seedMnemonic);
```
Just to be clear - this code is only called the very first time the user creates xud seed/wallet. If I'm reading the `ConnextClient` code correctly then it needs to call `initWallet` for every start. Is that correct?

If that's the case then we'll need to do changes our end to provide a seed for `ConnextClient`.

kilrau replied:
We actually just discussed this internally. You probably explained this before, but could you remind us again:
a) which keys exactly are derived from this seed, if more than one
b) how is the key/keys used
c) why does it need a new seed on every start

sangaman replied:
Am I correct in understanding that there'll be a mnemonic & seed associated with each channel in Connext? We definitely wouldn't want to use our "master" seed which is tied to on-chain funds as is currently being done.

Would it make sense to store connext channel seeds in the xud database and pass them to connext each time xud starts up?

Karl is right that the current implementation would not be able to initialize connext, as it would only call `initWallet` on the very first time we start xud.

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,513 @@
+/**
+ * A class representing a client to interact with connext.
+ */
+class ConnextClient extends SwapClient {
```
I think this is still missing the code where we listen for incoming payments with unknown preimages to request them from xud?

pedrouid committed:
rename channel to client on ConnextClient + remove unused variables

pedrouid committed:
remove network config from ConnextClient and url options required

sangaman commented on lib/Config.ts:
```diff
@@ -50,12 +51,13 @@ class Config {
   public network: XuNetwork;
-  public rpc: { disable: boolean, host: string, port: number };
-  public http: { host: string, port: number };
+  public rpc: { disable: boolean; host: string; port: number };
```
As a matter of style, I'd prefer to keep things as is with regards to member delimiters here. It actually seems that [semicolons are preferred in typescript](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/member-delimiter-style.md) but I'd rather keep styling consistent and not change this as part of the PR.

Enforcing this as a linting rule looks like it'd require migrating from tslint to typescript-eslint, which I started working on but is not a trivial effort, so for now it'd be good to just leave these lines as is.

sangaman commented on lib/Config.ts:
```diff
@@ -147,7 +157,14 @@ class Config {
       disable: false,
       certpath: path.join(lndDefaultDatadir, 'tls.cert'),
-      macaroonpath: path.join(lndDefaultDatadir, 'data', 'chain', 'bitcoin', this.network, 'admin.macaroon'),
+      macaroonpath: path.join(
```
This sort of style change is less significant as it doesn't break consistency, but as @erkarl said it makes the PR harder to review and I'd prefer to limit code churn and keep any changes in this PR relevant to its purpose.

pedrouid replied:
Totally, these were made before I disabled Prettier for this workspace... I will revert them

sangaman commented on lib/Config.ts:
```diff
@@ -168,20 +192,28 @@ class Config {
     try {
       configText = await fs.readFile(configPath, 'utf8');
-    } catch (err) {}
+    } catch (err) { }
```
Same here for unnecessary style changes.

sangaman commented on lib/Config.ts:
```diff
@@ -168,20 +192,28 @@ class Config {
         configProps = toml.parse(configText);
       } catch (e) {
-        throw new Error(`Error parsing config file at ${configPath} on line ${e.line}, column ${e.column}: ${e.message}`);
+        throw new Error(
```
And here.

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+  }
+
+  public tradingLimits = async (): Promise<TradingLimits> => {
+    // assume MAX_AMOUNT since Connext will re-collaterize accordingly
```
Can you explain a bit here what you mean by re-collaterizing? Eventually we'll want a way to know what our capacity is for sending and receiving payments with connext.

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+ */
+class ConnextClient extends SwapClient {
+  public readonly type = SwapClientType.Connext;
+  public readonly finalLock = 100;
```
Is this just a placeholder value or is this fixed with connext?

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+   * Initiates wallet for the Connext client
+   */
+  public initWallet = async (seedMnemonic: string[]) => {
+    this.wallet = new ConnextWallet(seedMnemonic.join(' '));
```
Following on the discussion about the mnemonic and wallet initialization, we wouldn't want to create the `ConnextWallet` here. As it's used with lnd, `initWallet` is only called the very first time we create a wallet within lnd. Rather, I'm thinking it makes sense to have the seed as part of the `ConnextClient` constructor.

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+      for (const [currency] of this.tokenAddresses) {
+        channelBalancePromises.push(this.channelBalance(currency));
+      }
+      await Promise.all(channelBalancePromises);
```
It looks like this is currently pulling back all the channel balances, but it doesn't do anything with the data it receives. Would it make sense to set this as our max inbound/outbound amounts?

sangaman commented on lib/connextclient/errors.ts:
```diff
@@ -0,0 +1,77 @@
+  CONNEXT_IS_UNAVAILABLE: codesPrefix.concat('.9'),
+  CONNEXT_HAS_NO_ACTIVE_CHANNELS: codesPrefix.concat('.10'),
+  CONNEXT_CLIENT_NOT_INITIATED: codesPrefix.concat('.11'),
+  CONNEXT_WALLET_NOT_INITIATED: codesPrefix.concat('.12'),
```
"Initialized" instead of "Initiated"?

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+      await client.isAvailable();
+
+      this.emit('connectionVerified', { newIdentifier: client.publicIdentifier });
+      this.setStatus(ClientStatus.ConnectionVerified);
```
This is a change I made recently, but I made a `setConnected` method that handles emitting the connected event and updating the status at once - in the future if there are any additional steps involved with updating status to connected we can just update that method. Can you use that here instead of these two lines?

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+
+      await client.isAvailable();
+
+      this.emit('connectionVerified', { newIdentifier: client.publicIdentifier });
```
Apologies if this wasn't clear, `newIdentifier` would only need to be set if it actually changed from its previous value. So it would be good to store the client identifier to a temp variable before verifying the connection, then only setting a value for `newIdentifier` if the new value doesn't match the old. You can use similar logic as is currently in LndClient.ts.

sangaman commented on lib/connextclient/types.ts:
```diff
@@ -0,0 +1,107 @@
+/**
+ * The payload for tokenPayment call.
+ */
+export type TokenPaymentRequest = {
```
This type and its members (with the `snake_case` variable names) are specific to Raiden and there's no reason they need to be reused for Connext. I'd say it's best that Connext types and properties/names align with the names used natively by the Connext client and/or standard javascript/typescript naming conventions (which usually doesn't use snake_case).

Edit: Same thing applies to other types in this class.

sangaman commented on lib/connextclient/errors.ts:
```diff
@@ -0,0 +1,77 @@
+   * or the amount is invalid, or there is no path to the target, or the identifier
+   * is already in use for a different payment."
+   */
+  INVALID: (message: string) => ({
```
It seems like most (or all?) of the errors from this point down aren't being used, but are rather carryovers from the Raiden REST HTTP error codes. No need to have them here, errors should be Connext specific and only included if they're actually being used.

pedrouid replied:
Correct, however they will be useful for using the new REST API client that Im working on

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+   * @param channel_address the address of the client to deposit to
+   * @param balance the amount to deposit to the client
+   */
+  public depositToChannel = async (
```
If there are no plans to use this method or no corresponding call within Connext, I think we can just remove it.

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+
+    if (response.secret) {
+      // remove '0x'
+      return response.secret.slice(2);
```
Is the '0x' included in the preimage response we'll get from Connext?

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+   * Returns the balances available in wallet for a specified currency.
+   */
+  public walletBalance = async (_currency?: string): Promise<WalletBalance> => {
+    return { totalBalance: 0, confirmedBalance: 0, unconfirmedBalance: 0 };
```
Do you expect that this method won't be relevant to Connext?

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,491 @@
+   */
+  public openChannel = async (): Promise<void> => {
+    if (!this.client) {
+      await this.initConnextClient();
```
I'm thinking rather than call this initialization logic, this may just not be relevant to connext (as with the closeChannel method) and can do nothing or throw an exception.

sangaman commented on lib/p2p/Peer.ts:
```diff
@@ -216,6 +216,9 @@ class Peer extends EventEmitter {
     if (clientType === SwapClientType.Raiden) {
       return this.nodeState.raidenAddress;
     }
+    if (clientType === SwapClientType.Connext) {
```
Might be nice to combine this and the if statements above into a switch statement, now that there are more than 2 possibilities.

sangaman commented on lib/p2p/Peer.ts:
```diff
@@ -252,20 +255,20 @@ class Peer extends EventEmitter {
+        retryConnecting?: boolean,
+        /** Port that Tor's exposed SOCKS5 proxy is listening on. */
+        torport: number,
+      }): Promise<packets.SessionInitPacket> => {
```
More style changes that I'd prefer to leave as-is.

sangaman reviewed (changes requested):
Thanks, there's a lot to process here but I think this is a good start.

Main takeaways:

- Keep style changes to existing code to a minimum to help further reviews and limit code churn. I'm looking at updating our linting setup - there are a good number of lines where this catches superfluous double whitespaces that I'd actually like to see fixed.

- There seems to be a decent amount of raiden carryover in types/errors/naming that isn't necessary.

- Wallet init and the mnemonic needs a different approach.

- We need a cleaner way to choose Connext over Raiden for tokens that are potentially supported by both, I'll look into handling this in a separate PR.

pedrouid committed:
add USDT

pedrouid committed:
revert style changes

pedrouid committed:
revert style changes

pedrouid committed:
revert style changes

pedrouid committed:
revert style changes

pedrouid committed:
rename intiated to intialized

pedrouid committed:
revert style changes

pedrouid committed:
use switch statement

pedrouid committed:
add USDT to limits and unit conversion

pedrouid committed:
add USDT pairs

pedrouid committed:
revert style changes

pedrouid committed:
remove snake case variables

pedrouid committed:
remove snake case variables

pedrouid committed:
refactor to rest api client

pedrouid committed:
tweak

pedrouid committed:
typo

pedrouid committed:
tweak

pedrouid committed:
Merge branch 'master' of https://github.com/ExchangeUnion/xud into connext

pedrouid committed:
update README

pedrouid committed:
linting

pedrouid committed:
updates

pedrouid committed:
update Transfer statuses

pedrouid committed:
update lookupPayment

pedrouid committed:
Merge branch 'master' of https://github.com/ExchangeUnion/xud into connext

pedrouid committed:
typo

pedrouid committed:
typo

Commit:
Merge branch 'master' into connext

Commit:
fix: add missing p2p proto definition

Commit:
chore: add connext sample-conf

Commit:
fix: init ConnextClient when swapclientmanager initializes

Commit:
fix: root rest-api endpoint

Commit:
fix: do not init ConnextClient with root seed

Commit:
fix: provide childSeed to ConnextClient upon every restart (noencrypt enabled)

Commit:
fix: filter tokenAddresses by ConnextClient SwapClientType

Commit:
fix: remove redundant logic from initSpecific, always initWallet

Commit:
fixup! fix: remove redundant logic from initSpecific, always initWallet

Commit:
fix: properly close ConnextClient for clean shutdown

Commit:
start HttpServer when ConnextClient is operational

Commit:
add /preimage endpoint to httpserver

Commit:
fix: active balance result

Commit:
subscribe to preimage; nodeInfo.connextAddress support

Commit:
fixup! subscribe to preimage; nodeInfo.connextAddress support

Commit:
listen for preimage - debug console.log

Commit:
throw error when sending taker payment

Commit:
add dummy getRoute

Commit:
exec transfer from taker side, listen for preimage

Commit:
connext taker happy flow working

Commit:
clear all subscriptions when xud starts

Commit:
maker happy flow working

Commit:
prevent duplicate preimage events

Commit:
remove temporary verifyConnection hack

Commit:
subscribe to CONDITIONAL_TRANSFER_RECEIVED

Commit:
fix CONDITIONAL_TRANSFER_RECEIVED_EVENT subscription

Commit:
provide seed and resubcribe within verifyConnection

Commit:
clear up openChannel

Commit:
cleanup comment

Commit:
emit htlcAccepted for non-eth erc20 assets

Commit:
reject invalid incoming transfer

Add logic to reject the incoming transfer if it does not match the
agreed upon hash, timelock, amount and tokenAddress.

Commit:
fixup! reject invalid incoming transfer

Commit:
remove todo from settleInovice

Commit:
working channel and wallet balance

Commit:
display connext info within getinfo call

Commit:
allow deposits via openchannel command

Commit:
Add README for local development

Commit:
unsubscribe before connecting

Commit:
disable unsubscribe all

Commit:
destructure new incoming transfer event properly

Commit:
fix incoming transfer currency/tokenAddress validation

Commit:
add dynamic currency for connext invoices

Commit:
store expected incoming transfers in memory

Commit:
move parseResponseBody under utils

Commit:
IncomingTransferRequest amount -> units

Commit:
TIMELOCK_BUFFER under ConnextClient

Commit:
ConnextPreimageRequest under connext/types

Commit:
ConnextIncomingTransferRequest under connext/types

Commit:
Move and rename TransferReceived and Preimage event type

Commit:
remove unused unsubscribeAll

Commit:
cleanup redundant errors

Commit:
cleanup redundant types

Commit:
remove redundant console.log

Commit:
remvoe arg checks TODOs

Commit:
openChannel formatting

Commit:
add debug log when accepting incoming transfers

Commit:
init Connext before Swaps (for recovery)

Commit:
do not throw when initializing connext

Commit:
fix ConnextPreimageRequest type

Commit:
init connext after pool

Commit:
remove TODOs

Commit:
use timelock buffer constant

Commit:
process preimage earlier

Commit:
add test token TOK support

Commit:
update readme with test ERC20 TOK

Commit:
add webhook host and port config option

sangaman commented on lib/connextclient/README.md:
````diff
@@ -0,0 +1,79 @@
+```
+### Running Connext node
+```bash
+make start
````
I needed to run `docker swarm init` at before this line would work.

erkarl replied:
Interesting... added it to the readme, just in case.

sangaman commented on lib/connextclient/README.md:
````diff
@@ -0,0 +1,79 @@
+```
+
+### Running Connext REST API client
+Create environment variables file and point to the Etheruem and Connext node
````
Maybe specify to create a `.env` file here as it wasn't immediately clear to me what to do here.

sangaman commented on lib/connextclient/README.md:
````diff
@@ -0,0 +1,79 @@
+
+```bash
+CONNEXT_ETH_PROVIDER_URL="http://0.0.0.0:8545"
+CONNEXT_NODE_URL="http://0.0.0.0:8080"
````
Will this conflict with the grpc web proxy which also uses this port by default?

sangaman reviewed (commented):
It looks like I have Connext up and running but I'm not seeing any balance, I'll reach out to you tomorrow @erkarl to make sure I'm not doing anything wrong with my setup.

I still need to review all the code changes, but I noticed that `bin/xud` needs to add the new connext config options so that we can pass them in as command line args.

Also this could use a rebase or merge with master, conflicts are minimal but it would be good to test with the latest xud code.

Commit:
Update README

Commit:
Remove undeployed USDT pair from simnet seeds

Commit:
Add TOK test ERC20 token to simnet seeds

Commit:
add TOK/ETH pair to simnet seeds

kilrau commented on lib/db/seeds/simnet.ts:
```diff
@@ -51,6 +51,7 @@ const pairs = [
   { baseCurrency: 'WETH', quoteCurrency: 'BTC' },
```
```suggestion
  { baseCurrency: 'WETH', quoteCurrency: 'BTC' },{ baseCurrency: 'ETH', quoteCurrency: 'BTC' }
```

Commit:
remove TIMELOCK_BUFFER from maker payment

raladev commented:
Mentions of the raiden in (which should not be if we remove it):
1. xucli help (xucli restore [backup_directory]   [raiden_database_path])
2. xucli getinfo (General XUD info - Raiden: Could not fetch status informatio)

Also:
1 Connecxt container is running instead of Ready in docker setup for status command (getinfo returns Ready state)
2. xud container too in Container is running state instead of Ready/Waiting, even after restart. logs does not contain errors or something that might attract attention(getinfo returns Not connected any peers in status field)

Commit:
set SwapCompleted phase after invoice settled

Commit:
only send SwapCompleted as taker

Commit:
do not set swap as complete from external packet

Commit:
Revert "only send SwapCompleted as taker"

This reverts commit 0a424aedf94a3f87e775433f6982d2b866117cf2.

Commit:
Revert "Revert "only send SwapCompleted as taker""

This reverts commit ca656a4b3f03d4e40924db6f45e31bb45e398946.

Commit:
upgrade changes

Commit:
remove SwapCompletedPacket

Commit:
remove docker swarm init from readme

erkarl commented:
Closing in favour of: https://github.com/ExchangeUnion/xud/pull/1424
